diff --git a/icosoc/common/firmware.c b/icosoc/common/firmware.c
index ea34dbc..72682a4 100644
--- a/icosoc/common/firmware.c
+++ b/icosoc/common/firmware.c
@@ -1,6 +1,8 @@
 #include <stdint.h>
 #include <stdbool.h>
 
+#define FLASH_CODE_MAX_SIZE_KB 256
+
 static void spiflash_begin()
 {
 	*(volatile uint32_t*)0x20000004 &= ~8;
@@ -175,7 +177,7 @@ int main()
 			spiflash_xfer(0);
 
 			uint8_t *p = (void*)(64*1024);
-			for (int i = 0; i < 128*1024; i++) {
+			for (int i = 0; i < FLASH_CODE_MAX_SIZE_KB * 1024; i++) {
 				if (i % 2048 == 0) console_putc('.');
 				p[i] = spiflash_xfer(0);
 			}
