diff --git a/icosoc/common/firmware.c b/icosoc/common/firmware.c
index ea34dbc..72682a4 100644
--- a/icosoc/common/firmware.c
+++ b/icosoc/common/firmware.c
@@ -1,6 +1,8 @@
 #include <stdint.h>
 #include <stdbool.h>
 
+#define FLASH_CODE_MAX_SIZE_KB 256
+
 static void spiflash_begin()
 {
 	*(volatile uint32_t*)0x20000004 &= ~8;
@@ -175,7 +177,7 @@ int main()
 			spiflash_xfer(0);
 
 			uint8_t *p = (void*)(64*1024);
-			for (int i = 0; i < 128*1024; i++) {
+			for (int i = 0; i < FLASH_CODE_MAX_SIZE_KB * 1024; i++) {
 				if (i % 2048 == 0) console_putc('.');
 				p[i] = spiflash_xfer(0);
 			}
diff --git a/icosoc/examples/riotos/Makefile b/icosoc/examples/riotos/Makefile
new file mode 100644
index 0000000..4d4d926
--- /dev/null
+++ b/icosoc/examples/riotos/Makefile
@@ -0,0 +1,23 @@
+CC = $(RISCV_TOOLS_PREFIX)gcc
+CFLAGS = --std=gnu99 -Os -MD -Wall
+OBJS = main.o icosoc.o
+
+default: help
+
+appimage.elf: $(OBJS)
+	$(CC) -T $(LDSCRIPT) -o appimage.elf $(OBJS)
+	chmod -x appimage.elf
+
+appimage.hex: appimage.elf
+	$(RISCV_TOOLS_PREFIX)objcopy -O verilog appimage.elf appimage.hex
+	chmod -x appimage.hex
+
+icosoc.mk: icosoc.cfg
+	../../icosoc.py
+
+clean::
+	rm -f appimage.hex appimage.elf $(OBJS) *.d
+
+include icosoc.mk
+-include *.d
+
diff --git a/icosoc/examples/riotos/icosoc.cfg b/icosoc/examples/riotos/icosoc.cfg
new file mode 100644
index 0000000..7260eef
--- /dev/null
+++ b/icosoc/examples/riotos/icosoc.cfg
@@ -0,0 +1,71 @@
+# icosoc for RIOTOS-icoboard
+board icoboard
+
+# enable compressed ISA support
+compressed_isa
+
+# custom interrupt vector address
+progaddr_irq 32'h0001_0080
+
+# enable memory mapped FLASH memory
+flashmem
+
+# UART serial - Digilent PmodUSBUART - PMOD3
+# http://store.digilentinc.com/pmodusbuart-usb-to-uart-interface/
+mod rs232 ser0
+  address 1
+  interrupt 4
+  param BAUD_RATE 115200
+  connect rx pmod3_3
+  connect tx pmod3_2
+
+# GPIO - PMOD1 and PMOD2
+mod gpio gpio0
+  address 2
+  connect IO pmod2 pmod1
+
+#########################
+#########################
+
+# Configuration for on-chip debugger
+debug_net resetn
+debug_net cpu_trap
+
+debug_net mem_valid
+debug_net mem_ready
+debug_net mem_instr
+
+debug_net mem_wstrb[3]
+debug_net mem_wstrb[2]
+debug_net mem_wstrb[1]
+debug_net mem_wstrb[0]
+
+debug_net mem_addr[31]
+debug_net mem_addr[30]
+debug_net mem_addr[29]
+debug_net mem_addr[28]
+debug_expr mem_addr_midbits 1 |mem_addr[27:18]
+debug_net mem_addr[17]
+debug_net mem_addr[16]
+debug_net mem_addr[15]
+debug_net mem_addr[14]
+debug_net mem_addr[13]
+debug_net mem_addr[12]
+debug_net mem_addr[11]
+debug_net mem_addr[10]
+debug_net mem_addr[9]
+debug_net mem_addr[8]
+debug_net mem_addr[7]
+debug_net mem_addr[6]
+debug_net mem_addr[5]
+debug_net mem_addr[4]
+debug_net mem_addr[3]
+debug_net mem_addr[2]
+debug_net mem_addr[1]
+debug_net mem_addr[0]
+
+debug_port pmod4_1 clk
+debug_port pmod4_2 resetn
+debug_port pmod4_3 cpu_trap
+debug_port pmod4_4 mem_valid
+
diff --git a/icosoc/icosoc.py b/icosoc/icosoc.py
index 737fdc4..3078ce8 100755
--- a/icosoc/icosoc.py
+++ b/icosoc/icosoc.py
@@ -37,6 +37,7 @@ enable_flashmem = False
 enable_flashpmem = False
 enable_noflashboot = False
 enable_fastflashboot = False
+progaddr_irq = "32'h 0000_0010"
 
 debug_depth = 256
 debug_trigat = 0
@@ -112,6 +113,7 @@ def parse_cfg(f):
     global enable_flashpmem
     global enable_noflashboot
     global enable_fastflashboot
+    global progaddr_irq
     global debug_code_append
 
     current_mod_name = None
@@ -184,6 +186,12 @@ def parse_cfg(f):
             enable_fastflashboot = True
             continue
 
+        if line[0] == "progaddr_irq":
+            assert len(line) == 2
+            assert current_mod_name is None
+            progaddr_irq = line[1]
+            continue
+
         if line[0] == "debug_net":
             assert len(line) == 2
             debug_signals[line[1]] = line[1]
@@ -480,6 +488,9 @@ icosoc_v["30-raspif"].append("""
     wire trigger_2;  // unused
     wire trigger_3;  // unused
 
+    reg recv_ep2_irq = 0;
+    reg recv_ep2_valid_q1;
+
     icosoc_raspif #(
         .NUM_RECV_EP(4),
         .NUM_SEND_EP(4),
@@ -571,7 +582,8 @@ icosoc_v["40-cpu"].append("""
         .COMPRESSED_ISA(<compisa>),
         .ENABLE_MUL(<muldiv>),
         .ENABLE_DIV(<muldiv>),
-        .ENABLE_IRQ(1)
+        .ENABLE_IRQ(1),
+        .PROGADDR_IRQ(<progaddr_irq>)
     ) cpu (
         .clk       (clk      ),
         .resetn    (resetn   ),
@@ -587,7 +599,8 @@ icosoc_v["40-cpu"].append("""
     );
 """
 .replace("<compisa>", ("1" if enable_compressed_isa else "0"))
-.replace("<muldiv>", ("1" if enable_muldiv_isa else "0")))
+.replace("<muldiv>", ("1" if enable_muldiv_isa else "0"))
+.replace("<progaddr_irq>", progaddr_irq))
 
 icosoc_v["50-mods"].append("""
     // -------------------------------
@@ -595,6 +608,8 @@ icosoc_v["50-mods"].append("""
 """)
 
 irq_terms = list()
+irq_terms.append("%s << %s" % ("recv_ep2_irq", 3))
+
 txt = icosoc_v["50-mods"]
 for m in mods.values():
     if m["addr"] is not None:
@@ -832,6 +847,11 @@ icosoc_v["72-bus"].append("""
         if (send_ep2_ready)
             send_ep2_valid <= 0;
 
+        recv_ep2_irq <= 0;
+        if (!recv_ep2_valid_q1 && recv_ep2_valid)
+            recv_ep2_irq <= 1;
+        recv_ep2_valid_q1 <= recv_ep2_valid;
+
         recv_ep2_ready <= 0;
 
         if (!resetn) begin
@@ -845,6 +865,8 @@ icosoc_v["72-bus"].append("""
 
             send_ep2_valid <= 0;
             spiflash_state <= 0;
+            recv_ep2_irq <= 0;
+            recv_ep2_valid_q1 <= 0;
         end else
         if (mem_valid && !mem_ready) begin
             (* parallel_case *)
diff --git a/icosoc/mod_rs232/mod_rs232.v b/icosoc/mod_rs232/mod_rs232.v
index f32be4d..c5567a0 100644
--- a/icosoc/mod_rs232/mod_rs232.v
+++ b/icosoc/mod_rs232/mod_rs232.v
@@ -8,6 +8,7 @@ module icosoc_mod_rs232 (
 	input [31:0] ctrl_wdat,
 	output reg [31:0] ctrl_rdat,
 	output reg ctrl_done,
+	output reg ctrl_irq,
 
 	inout tx, rx
 );
@@ -55,6 +56,7 @@ module icosoc_mod_rs232 (
 	always @(posedge clk) begin
 		rxd_q <= rxd;
 		recv_shift_in <= 0;
+		ctrl_irq <= 0;
 
 		if (!resetn) begin
 			rx_state <= 0;
@@ -76,6 +78,7 @@ module icosoc_mod_rs232 (
 			if (rx_state == 8) begin
 				recv_shift_in <= 1;
 				rx_state <= 0;
+				ctrl_irq <= 1;
 			end
 		end
 	end
